### ISSUE TITLE
Any tips for MVP in Swift?

### ISSUE THREAD

**Issue by andreif** [kyleneideck/BackgroundMusic#440]
Hi, could you please advise on how to implement a basic function for forwarding output to input in Swift? I don't really need any fancy features so some minimal MVP would be just fine. You don't need to be specific if it feels to hard, just some main steps in making it possible. Why? I am having trust issues in installing 3rd party software, even open source if it's c++ & friends. Also, it would be a nice learning experience since Swift makes it so much simpler and efficient.

**Comment by andreif** [comment]
Is it what driver is for? https://github.com/kyleneideck/BackgroundMusic/tree/master/BGMDriver/BGMDriver

**Comment by kyleneideck** [comment]
>Is it what driver is for? https://github.com/kyleneideck/BackgroundMusic/tree/master/BGMDriver/BGMDriver

Yeah, pretty much. The main function of BGMDriver is to be a "loopback" driver. It publishes a virtual input device and a virtual output device. When apps play audio to the output device, BGMDriver repeats that audio on the input device.

The driver interface it implements is [AudioServerPlugIn](https://github.com/phracker/MacOSX-SDKs/blob/master/MacOSX10.15.sdk/System/Library/Frameworks/CoreAudio.framework/Versions/A/Headers/AudioServerPlugIn.h), so that's probably where you'd want to start. I'm not sure whether Swift would be OK to use for the IO functions, since they have to be realtime-safe, but Swift uses ARC (as opposed to a less-predictable garbage collection algorithm), so I think that means it could be fine.

I actually wrote something like the "playthrough" functionality from BGMApp in Swift for a different project. It's a similar concept, except it's forwarding the audio from the virtual input device to the real output device.

IIRC I tried using `AVAudioEngine` from `AVFoundation` and then tried using the AUHAL audio unit. I don't remember exactly why those approaches didn't work out, but I didn't spend much time on them. I think it might have been that they were causing problems with other parts of the project somehow. In the end I used Apple's [CAPlayThrough](https://developer.apple.com/library/archive/samplecode/CAPlayThrough/Introduction/Intro.html) sample code, but that's written in C++ as well, so it might not be that useful to you.
